File: [0300][Medium][Longest_Increasing_Subsequence]/Longest_Increasing_Subsequence_3.py
# leetcode time     cost : 84 ms
# leetcode memory   cost : 13.8 MB 
# Time  Complexity: O(n**2)
# Space Complexity: O(n)
# slolution 3, DP,
class Solution(object):
    def lengthOfLIS(self, nums):
        if not nums:  return 0
        best = [nums[0]]
        for i in range(1, len(nums)):
            index = self.upperBound(best, nums[i], 0, len(best))
            if index == len(best): best.append(nums[i])
            else: best[index] = nums[i]
        return len(best)

    # Return smallest index i such that array[i] >= target
    # or return end+1 if none (target > all array elements)
    def upperBound(self, array, target, start, end):
        if start >= end: return start
        m = (end+start)//2
        if array[m] >= target:
            if m == start or array[m-1] < target: return m
            return self.upperBound(array, target, start, m)
        return self.upperBound(array, target, m+1, end)

File: [0300][Medium][Longest_Increasing_Subsequence]/Longest_Increasing_Subsequence_3.java
// leetcode time     cost : 42 ms
// leetcode memory   cost : 38.7 MB 
// Time  Complexity: O(n**2)
// Space Complexity: O(n)
// slolution 3, DP,
public class Solution {
    public int lengthOfLIS(int[] nums) {
      if (nums.length == 0) {
          return 0;
      }
      int maxLength = 1;
      int[] dp = new int[nums.length+1];
      // init dp[i], dp[i] is the max LISequece length which contains the nums[i] when come to nums[i],
      for (int i = 0; i < nums.length; ++i) 
          dp[i] = 1;
      for (int i = 1; i < nums.length; ++i) {
          for (int j = 0; j < i; ++j) {
              if (nums[j] < nums[i]) {
                  dp[i] = Math.max(dp[i], dp[j] + 1) ;
              }
          }
          maxLength = Math.max(maxLength, dp[i]);
      }
      return maxLength;
  }  
  // generate the subsequence to maintain its max length.

  public int lengthOfLIS_1(int[] nums) {
      if (nums.length == 0) {
          return 0;
      }
      // dp[i] is the longest increasing subsequence when come to nums[i]
      List<Integer> dp = new ArrayList<>();
      dp.add(nums[0]);
      int maxIndex = 0;
      for (int i = 0; i < nums.length; i++) {
            if (dp.get(maxIndex) < nums[i]) {
                dp.add(nums[i]);
                maxIndex++;
                
            }
            else{
                int tmpIndex = 0;
                //time can be saved if use binary search in while since dp is a increasing sequence.
                //refer to solution 4.
                while(dp.get(tmpIndex) < nums[i])
                    tmpIndex++;
                dp.set(tmpIndex,nums[i]);
            }
            //System.out.println("dp is :" + dp);
      }
      return dp.size();
  }
}

public class Longest_Increasing_Subsequence_3 {
    public static void main(String args[]) {
        int[] nums = {1,3,6,7,9,4,10,5,6}  ;   // expect is 6
        Solution Solution_obj = new Solution();
        int result = Solution_obj.lengthOfLIS(nums);
        System.out.println("In java code,return value is :" + (result));
    }
}